### ISSUE TITLE
confidence reporting for PerformanceNavigationTiming

### ISSUE THREAD

**Issue by mwjacksonmsft** [w3ctag/design-reviews#878]
I'm requesting a TAG review of systemEntropy addition to PerformanceNavigationTiming.

This proposal adds a new ‘systemEntropy’ field to the PerformanceNavigationTiming struct enabling developers to discern if the page load occurs during a non-optimal performance state.

[Update: 2025-05: The name of this feature is now "confidence", rather than "systemEntropy"]

  - Explainer¹ (minimally containing user needs and example code): [url](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/PerformanceNavigationTiming%20for%20User%20Agent%20Launch/explainer.md)
  - Specification URL: N/A
  - Tests: not yet
  - User research: N/A
  - Security and Privacy self-review²: [url](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/PerformanceNavigationTiming%20for%20User%20Agent%20Launch/security-privacy.md)
  - GitHub repo (if you prefer feedback filed there): N/A
  - Primary contacts (and their relationship to the specification):
      - Mike Jackson (mwjacksonmsft), Microsoft
  - Organization(s)/project(s) driving the specification: Microsoft
  - Key pieces of existing multi-stakeholder review or discussion of this specification: N/A
  - External status/issue trackers for this specification (publicly visible, e.g. Chrome Status): https://chromestatus.com/feature/5186950448283648

Further details:

  - [x] I have reviewed the TAG's [Web Platform Design Principles](https://www.w3.org/TR/design-principles/)
  - Relevant time constraints or deadlines: [please provide]
  - The group where the work on this specification is currently being done: W3C WebPerf WG
  - The group where standardization of this work is intended to be done (if current group is a community group or other incubation venue): W3C WebPerf WG
  - Major unresolved issues with or opposition to this specification: N/A
  - This work is being funded by: Microsoft

You should also know that...

[please tell us anything you think is relevant to this review]

We'd prefer the TAG provide feedback as (please delete all but the desired option):
  💬 leave review feedback as a **comment in this issue** and @-notify [mwjacksonmsft]

**Comment by plinss** [comment]
We haven't had a chance to dive into this thoroughly yet, but first impressions:
* The term "entropy" has a number of other connotations and may not be the best term here. Possibly something like "systemLoad" might be more obvious to users.
* I don't have specific examples, but this leads to concern about possibly introducing additional information for side-channel attacks or user fingerprinting (see battery status API). I accept that this is past data, but it's strongly correlated with high resolution timers. Have crypto and privacy experts evaluated this aspect?

**Comment by mwjacksonmsft** [comment]
Thanks for the feedback. 

- I'm open to feedback on the name. I have a concern that 'load' implies some system resource usage is high, which may or may not be the case, especially during cold start where there could be lock contention, or delays while loading binaries off disk. 

- Is the concern that a third-party script is included on siteA and siteB, that the script might be able to create a short-term identifier to track the user across the two sites?

**Comment by mwjacksonmsft** [comment]
My apologies for the delayed reply. I'm working with the WebPerfWG to ensure that this information is exposed in a way that conforms with the privacy principles outlined here: https://docs.google.com/presentation/d/19TOz4mXRsYt8tkqzH8io_BrYiZBXhqGyD646gJy-x6I/edit#slide=id.p

**Comment by hadleybeeman** [comment]
Thanks, @mwjacksonmsft — we'll wait to hear back from you then.

**Comment by mwjacksonmsft** [comment]
Hi Tag - 

We've iterated on the design of this API in the WebPerfWG. The explainer, and Chrome status entry pages have all been updated. Updated details are captured in the explainer, but the high-level summary is that the field has been renamed to 'confidence', and noise via a randomized response algorithm has been introduced to reduce fingerprinting risks. 

The explainer also gives an example of how a RUM provider might debias the aggregate data.

**Comment by martinthomson** [comment]
Looking at this, I wonder if there isn't a far simpler solution:

* If the system is loaded (i.e., those cases where confidence would be "low" in the proposed design), the API might fail.
* To provide sites with some baseline failure rate so that failures are not directly correlated with heavily loaded clients, have some sort of non-trivial, but random rate of failure.  5% might do for that.  The only purpose of that is to establish that failures are completely normal and there is no expectation that any given visit will result in a report.  It might even be possible to go higher.

Not submitting data like that would also be more private than what is proposed.  Because it doesn't reveal timing information when the browser is loaded.  That is likely a situation where more reports are sent out as multiple windows are loaded at the same time, which means that correlation between low confidence reports might reveal good cross-site correlation information, even with randomized response.

The consequence of that would be that you would generally only get performance metrics from browsers that are not under stress.  The risk is that you might incentivize manual data collection methods if you don't provide metrics.  After all, most of the underlying factors are observable anyway, but my guess is that most sites wouldn't bother to collect noisy metrics like that.  After all, I'd guess that most page loads are done on machines that aren't in this heavily loaded state.

This came from me looking at the "differential privacy" solution being proposed.  In doing so, I noticed that the approach lacked the necessary sensitivity analysis.  It didn't identify the privacy unit, but I'd assume that it was the report, where the real privacy risk comes from a privacy unit with a larger scope, like the user or user plus site.  Covering each page load doesn't help if your noise can be averaged out by having multiple reports generated.  (In other words, I don't think that claims about this providing differential privacy are good without more analysis.)

**Comment by mwjacksonmsft** [comment]
Hi @martinthomson - 

Thank you for the reply, and your patience with my delayed response.

I'm concerned that if we allow the API to fail (or return an empty list), we will end up causing web interop issues. A quick search on GitHub reveals that many developers assume that an entry is always returned: https://github.com/search?q=performance.getEntriesByType%28%27navigation%27%29%5B0%5D.&type=code

Additionally, samples of how to use the API also exclude empty list/null checks:
 * https://web.dev/articles/navigation-and-resource-timing
 * https://w3c.github.io/navigation-timing/#example-3

If we were able to address those interop concerns, I'd be concerned that it would be less private in the "high" confidence case. The goal of this work is to provide developers and RUM providers a signal indicating that the metrics might not accurately represent the website’s performance due to external factors beyond their control. Improving the privacy of existing metrics returned by this API is not within the scope of this work. The noise being added is to prevent the confidence bit from serving as an additional entropy source for user identification.

**Comment by martinthomson** [comment]
As is often the case, I was focusing too narrowly. Let's put this all in perspective.

The fundamental problem with the underlying API is its usefulness for fingerprinting.  With good performance metrics, you are able to segregate a user population by various performance characteristics: the speed at which connections are created, the speed at which they download content, the speed that they perform layout, and the speed that they execute JavaScript.

A site visitor in the slowest 10% of visitors to one site is probably also in the slowest 10% of visitors to another site.  (That the spec has [a privacy considerations section](https://w3c.github.io/navigation-timing/#privacy) that fails to acknowledge this is a problem.)

Much of this is already observable on the platform, but the API makes these things trivial to measure as a side-effect of loading content. The performance timing API therefore takes what might otherwise require active fingerprinting and serves up the information with no special effort required.

The main thing that stops this from being a reliable source of fingerprinting information is the sheer amount of variability in the metrics.  That variability is what allows us to pretend "this is fine".  As long as we're in a fingerprinting-rich environment - or we have browsers with third party cookies - maybe we can maintain that pretense.

What is unique with the addition is that "confidence" is reflecting a value that doesn't have the inherent noise of all the rest of the timing information.

Making the "confidence" parameter noisy only makes it necessary to collect multiple samples (epsilon values increase over multiple observations, eliminating noise).  It seems to me like you want to encourage sites to take multiple observations across a user population rather than of a single user, but there are no inbuilt systems that prevent abuse.

Either way, it looks like the problem is inherent in the underlying API, less so the proposed addition.  But if the underlying API is bad for privacy, can we endorse extending it further to make it worse?

**Comment by yoavweiss** [comment]
> Either way, it looks like the problem is inherent in the underlying API

Just to make sure I understand what you are saying - are you suggesting that NavigationTiming (and performance measurements in general) are a fingerprinting vector?

**Comment by martinthomson** [comment]
> are you suggesting that NavigationTiming (and performance measurements in general) are a fingerprinting vector?

Yep.  I'm not saying that they aren't worth it, but the addition of this proposal highlights that we haven't done a lot to protect these.  That might still be the right answer, but it would be good if that were a deliberate choice.

**Comment by yoavweiss** [comment]
We discussed this on the TAG call and we don't have concerns with this particular API. Its use of differential privacy explicitly avoids exposing information about any particular user, which is great.

One point which may benefit web developers would be if the spec mandated or recommended a certain epsilon value across different implementations, and that would allow developers to debias their metrics across different implementations. With that said, the fact that the value is reflected using randomizedTriggerRate would still enable developers to handle differences between implementations if those arise.

At the same time, there's some concern related to fingerprinting and web performance APIs in general. Slow users (through various dimensions of slowness: network latency, bandwidth or CPU) are likely to be slow across different sites, which can contribute fingerprinting bits to would allow for their persistent re-identification across sites.
This information is available through [ancillary APIs](https://www.w3.org/TR/privacy-principles/#dfn-ancillary-apis-computed-from-existing-information) and [non-ancillary APIs](https://www.w3.org/TR/privacy-principles/#dfn-non-ancillary-apis) alike, and hence it seems unlikely that the situation can be improved. At the same time, it could be good to mention that concern in the relevant APIs' Privacy and Security sections.
